src/metrics/mod.rs

```diff
@@ -67,6 +78,64 @@ metrics! {
     namespace: "docsrs",
 }
 
+#[derive(Debug, Default)]
+pub(crate) struct RecentlyAccessedReleases {
+    crates: DashMap<i32, Instant>,
+    versions: DashMap<i32, Instant>,
+    platforms: DashMap<(i32, TargetAtom), Instant>,
+}
+
+impl RecentlyAccessedReleases {
+    pub(crate) fn new() -> Self {
+        Self::default()
+    }
+
+    pub(crate) fn record(&self, krate: i32, version: i32, target: &str) {
+        self.crates.insert(krate, Instant::now());
```

Comment:
Should this have a cap on the number of crates recorded? Right now it will grow without bound and only evict entries each hour, which seems like it would be really easy to DOS.

On a related point, is it possible to do all this calculation on the grafana side? That way we'd only have to hold on to the entries long enough to send them to `metrics`, and could delete them immediately after.

Reply:
It evicts entries every time metrics are requested, if they're an hour old. But I can probably throw some massive limit on there just in case (say 64 bytes per entry (4 byte key, 8 byte hash, 16 byte value, padding, overhead), 100,000 entries = 6.4M, times two for hashmap overhead, times two for margin = 25.6M per map; I'm less sure about what that would do to the metrics response times going through all those entries every time, so maybe 10,000 as a more conservative limit; some very quick benchmarking on my machine shows perfect linear scaling and ~11ms for 100,000 entries in all 3 maps).

I have no idea how you'd do it in grafana, it doesn't fit into any of the prometheus metric types directly. (If we had logs going into kibana, then I'd have some idea about how to pull this data from there instead of tracking it in service).

Reply:
75 MB seems like more than enough and won't affect our memory usage too much, let's go with that. Ideally it would be a LRU cache but if that's hard to do I think it's fine to just refuse to record more entries.